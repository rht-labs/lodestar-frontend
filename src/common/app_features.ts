// All roles defined by the backend to be used in the frontend should be enumerated in this object.

export const APP_FEATURES = {
  admin: 'admin',
  writer: 'writer',
  reader: 'reader',
  notifications: 'notifications',
  engagementCardIcons: 'engagementCardIcons',
  resetUser: 'resetUser',
};

export type AppFeature = keyof typeof APP_FEATURES;
