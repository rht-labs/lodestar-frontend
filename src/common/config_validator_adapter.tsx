import { ValidatorFactory } from '../schemas/validators';
import { EngagementFormConfig } from '../schemas/engagement_config';

export const getValidatorsFromFormOptions = (
  formOptions: EngagementFormConfig = {}
) =>
  Object.keys(formOptions || {}).reduce(
    (previousSchemaGroups, currentGroupKey) => {
      return {
        ...previousSchemaGroups,
        ...Object.keys(formOptions[currentGroupKey] ?? {}).reduce(
          (previousGroupFields, currentFieldKey) => ({
            ...previousGroupFields,
            [currentFieldKey]: (
              formOptions?.[currentGroupKey]?.[currentFieldKey]?.validators ||
              []
            ).map(ValidatorFactory),
          }),
          {}
        ),
      };
    },
    {}
  );
