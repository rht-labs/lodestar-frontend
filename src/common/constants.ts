export const APPLICATION_CONTENT_CONTAINER_ID = 'main-application-content';
