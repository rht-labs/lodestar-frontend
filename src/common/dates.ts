import { format as formatDate, utcToZonedTime } from 'date-fns-tz';
export const formatUtcDate = (date?: Date): string | undefined => {
  if (!date) {
    return '';
  }
  const dateString = formatDate(utcToZonedTime(date, 'UTC'), 'MMM dd, yyyy', {
    timeZone: 'UTC',
  });
  return dateString;
};
