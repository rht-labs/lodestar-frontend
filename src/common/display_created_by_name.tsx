import React from 'react';

export function DisplayCreatedByName({
  userFromServer,
  lastUpdatedBy,
}: {
  userFromServer?: string;
  lastUpdatedBy?: string;
}) {
  return <>{!!userFromServer ? userFromServer : lastUpdatedBy}</>;
}
