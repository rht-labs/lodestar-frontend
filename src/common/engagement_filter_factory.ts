import {
  EngagementFilter,
  EngagementSortFields,
} from '../schemas/engagement_filter';
import { Engagement, getEngagementStatus } from '../schemas/engagement';

function hasSearchTermMatch(engagement: Engagement, searchTerm: string = '') {
  if (!searchTerm || !engagement) {
    return true;
  }
  const cleanedSearchTerm = searchTerm?.trim()?.toLowerCase();
  const cleanedCustomerName = engagement.customer_name.toLowerCase().trim();
  const cleanedProjectName = engagement.project_name.toLowerCase().trim();
  return (
    cleanedProjectName.includes(cleanedSearchTerm) ||
    cleanedCustomerName.includes(cleanedSearchTerm)
  );
}

/**
 * The Engagement Filter Factory accepts a filter definition as a parameter.
 * It parses the filter definition and returns a single filter function that
 * filters an array of Engagements based on the terms of the filter definition.
 */
export const engagementFilterFactory = (
  filter?: EngagementFilter
): ((engagement: Engagement) => boolean) => {
  if (!filter || Object.keys(filter).length === 0) {
    return () => true;
  }
  return (engagement: Engagement) => {
    const filterTestResults: boolean[] = [];
    if (filter.allowedStatuses) {
      filterTestResults.push(
        filter.allowedStatuses.includes(getEngagementStatus(engagement))
      );
    }
    if (filter.searchTerm) {
      filterTestResults.push(
        hasSearchTermMatch(engagement, filter?.searchTerm)
      );
    }
    /**
     * If filterTestResult has a length of 0, this means no tests were performed.
     * The default behavior of the filter should be to show engagements.
     * Therefore, we return true
     */
    if (filterTestResults.length === 0) {
      return true;
    }
    /**
     * If filterTestResults has results, we only return if every result is true.
     */
    return filterTestResults.every(r => !!r);
  };
};

const engagementSorters = (sortOrderMultiplier: number = 1) => {
  return {
    startDate: (a: Engagement, b: Engagement) =>
      (a.start_date > b.start_date ? 1 : -1) * sortOrderMultiplier,

    endDate: (a: Engagement, b: Engagement) =>
      (a.end_date > b.end_date ? 1 : -1) * sortOrderMultiplier,

    customerName: (a: Engagement, b: Engagement) =>
      (a.customer_name > b.customer_name ? 1 : -1) * sortOrderMultiplier,

    projectName: (a: Engagement, b: Engagement) =>
      (a.project_name > b.project_name ? 1 : -1) * sortOrderMultiplier,
  };
};

/**
 * The Engagement Sort Factory parses the `sort` field of the Engagement filter definition.
 * It returns a single function that can be passed to the `sort` method of an Engagement array.
 */
export const engagementSortFactory = ({ sort }: EngagementFilter = {}): ((
  a: Engagement,
  b: Engagement
) => number) => {
  /**
   * The sort order multiplier translates the isAscending boolean field into
   * an integer that can invert the sort order of the array.
   */
  const sortOrderMultiplier =
    sort?.isAscending === undefined || sort?.isAscending ? 1 : -1;
  if (sort?.sortField === EngagementSortFields.startDate) {
    return engagementSorters(sortOrderMultiplier).startDate;
  }
  if (sort?.sortField === EngagementSortFields.endDate) {
    return engagementSorters(sortOrderMultiplier).endDate;
  }
  if (sort?.sortField === EngagementSortFields.customerName) {
    return engagementSorters(sortOrderMultiplier).customerName;
  }
  if (sort?.sortField === EngagementSortFields.projectName) {
    return engagementSorters(sortOrderMultiplier).projectName;
  }
  return () => 0;
};
