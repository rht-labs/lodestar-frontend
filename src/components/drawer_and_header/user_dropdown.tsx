import React, { useState } from 'react';
import { useSession } from '../../context/auth_context/auth_context';
import { Dropdown, DropdownItem, DropdownToggle } from '@patternfly/react-core';
import { Link } from 'react-router-dom';
import { UserProfile } from '../../schemas/user_profile';

export interface UserDropdown {}

export function UserDropdown(props: UserDropdown) {
  const authContext = useSession();
  const [isOpen, setIsOpen] = useState(false);
  const dropdownItems = [
    <DropdownItem value="logOut" key="logout" component={Link} to="/logout">
      Log Out
    </DropdownItem>,
  ];

  const onSelect = (_?: React.SyntheticEvent<HTMLDivElement>) => {
    setIsOpen(false);
  };

  return (
    <div
      style={{
        width: 'auto',
        height: 50,
        justifyContent: 'center',
        display: 'flex',
        alignItems: 'center',
      }}
    >
      <Dropdown
        isPlain={true}
        dropdownItems={dropdownItems}
        toggle={
          <DropdownToggle onToggle={() => setIsOpen(!isOpen)}>
            {UserProfile.getDisplayName(authContext?.sessionData?.profile)}
          </DropdownToggle>
        }
        isOpen={isOpen}
        onSelect={onSelect}
      />
    </div>
  );
}
