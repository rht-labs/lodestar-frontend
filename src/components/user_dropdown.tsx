import React, { useContext, useState } from 'react';
import { SessionContext } from '../context/session_context';
import { Dropdown, DropdownToggle, DropdownItem } from '@patternfly/react-core';
import { Link } from 'react-router-dom';
export interface UserDropdown {}
export function UserDropdown(props: UserDropdown) {
  const sessionContext = useContext(SessionContext);
  const [isOpen, setIsOpen] = useState(false);
  const dropdownItems = [
    <DropdownItem value="logOut" key="logout">
      <Link to="/logout">Log Out</Link>
    </DropdownItem>,
  ];

  const onSelect = (_: React.SyntheticEvent<HTMLDivElement>) => {
    setIsOpen(false);
  };

  return (
    <div
      style={{
        width: 'auto',
        height: 50,
        justifyContent: 'center',
        display: 'flex',
        alignItems: 'center',
      }}
    >
      <Dropdown
        isPlain={true}
        dropdownItems={dropdownItems}
        toggle={
          <DropdownToggle onToggle={() => setIsOpen(!isOpen)}>
            Hello, {sessionContext.profile?.displayName}
          </DropdownToggle>
        }
        isOpen={isOpen}
        onSelect={onSelect}
      ></Dropdown>
    </div>
  );
}
