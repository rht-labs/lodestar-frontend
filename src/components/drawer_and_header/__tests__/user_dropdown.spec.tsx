import React from 'react';
import { UserDropdown } from '../user_dropdown';
import { render } from '@testing-library/react';
import { TestStateWrapper } from '../../../common/test_state_wrapper';
import { act } from 'react-dom/test-utils';

describe('User Dropdown UI', () => {
  test('should match snapshot', async () => {
    await act(async () => {
      const rendered = render(
        <TestStateWrapper>
          <UserDropdown />
        </TestStateWrapper>
      );
      expect(rendered).toMatchSnapshot();
    });
  });
});
