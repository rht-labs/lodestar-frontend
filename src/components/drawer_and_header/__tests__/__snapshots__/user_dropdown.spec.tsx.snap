// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Dropdown UI should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="width: auto; height: 50px; justify-content: center; display: flex; align-items: center;"
      >
        <div
          class="pf-c-dropdown"
          data-ouia-component-id="0"
          data-ouia-component-type="PF4/Dropdown"
          data-ouia-safe="true"
        >
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="pf-c-dropdown__toggle pf-m-plain"
            id="pf-dropdown-toggle-id-0"
            type="button"
          >
            
            <span
              class="pf-c-dropdown__toggle-icon"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  transform=""
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="width: auto; height: 50px; justify-content: center; display: flex; align-items: center;"
    >
      <div
        class="pf-c-dropdown"
        data-ouia-component-id="0"
        data-ouia-component-type="PF4/Dropdown"
        data-ouia-safe="true"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="pf-c-dropdown__toggle pf-m-plain"
          id="pf-dropdown-toggle-id-0"
          type="button"
        >
          
          <span
            class="pf-c-dropdown__toggle-icon"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                transform=""
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
